### PR TITLE
Annotation legend

### LINKED ISSUE #8236: Legend does not show 'annotate'
When using `annotate` to draw an arrow, it does not show up in the legend.

Example code:
```python
import matplotlib.pylab as plt
%matplotlib inline

fig, ax = plt.subplots(1)
ax.plot([0, 1], [0, 0], label='line1')
ax.plot([0, 1], [1, 1], label='line2')
ax.annotate("",
            xy=(0.4,1.0),
            xytext=(0.4,0.0),
            arrowprops={'arrowstyle':'<->', 'color':'C7' },
            label='distance')
ax.legend()
```
gives:
annotate

I would expect an entry "distance" in the legend, with a grey double-sided arrow as symbol.

(I use Python 3.6.0 and Matplotlib 2.0.0 from Archlinux repository)

### PR DESCRIPTION
**Pull request by jlecoeur**
Fixes #8236 

I added new legend handlers for Text, FancyArrowPatch and Annotation artists (HandlerText, HandlerFancyArrowPatch and HandlerAnnotation).

HandlerAnnotation can draw different artists in the legend:
- if the annotation has no text and no arrow, it draws a white Rectangle
- if the annotation has only text, it uses HandlerText to draw a text. The text is replaced by 'Aa' if it is longer than one character. (This is configurable by setting rep_str and rep_maxlen.)
- if the annotation has only an arrow, it uses HandlerFancyArrowPatch to draw an arrow
- If the annotation has both text and arrow, it uses HandlerTuple to draw both.

In HandlerTuple, I added the possibility of different width ratios for the artists. HandlerAnnotation uses a width_ratios=[1,4] (ie. the arrow is 4 times wider than the text).

Example code:
```python

import matplotlib
import matplotlib.pylab as plt
from matplotlib.patches import FancyArrowPatch
from matplotlib.legend_handler import HandlerAnnotation

fig, ax = plt.subplots(1)
ax.plot([0, 1], [0, 0], label='line1')
ax.plot([0, 1], [1, 1], label='line2')

# no text, no arrow
ax.annotate("",
            xy=(0.1,0.5),
            xytext=(0.1,0.5),
            label='annotation (empty)')

# text, no arrow
my_annotation = ax.annotate("X",
            xy=(0.1,0.5),
            xytext=(0.1,0.5),
            color='C2',
            label='annotation (text, no arrow)')

# no text, arrow
ax.annotate("",
            xy=(0.3,1.0),
            xytext=(0.3,0.0),
            arrowprops={'arrowstyle':'<->', 'color':'C7' },
            label='annotation (no text, arrow)')

# text, arrow
ax.annotate("Some long string",
            xy=(0.4,1.0),
            xytext=(0.35,0.1),
            arrowprops={'arrowstyle':'->', 'color':'C2' },
            color='C1',
            label='annotation (text + arrow)')

# Fancy arrow patch
arrpatch = FancyArrowPatch([0.5, 0.8], [0.9, 0.9],
                           arrowstyle='<|-',
                           mutation_scale=20,
                           color='C3',
                           label='arrowpatch')
ax.add_patch(arrpatch)

# Long text, will not be used in legend
ax.text(x=0.1, y=0.1,
        s='Hello',
        color='C5',
        label='text')

# Short text, copied in legend
ax.text(x=0.1, y=0.2,
        s='Z',
        color='C0',
        label='short text')

ax.legend(handler_map={my_annotation: HandlerAnnotation(rep_str='Abcde', rep_maxlen=0)})
fig.show()
```
output:
annotate

### CODE REVIEW

File: lib/matplotlib/legend_handler.py [hunk 1]
```diff
@@ -661,3 +692,59 @@ def create_artists(self, legend, orig_handle,
...
+        self.update_prop(t, orig_handle, legend)
+        t.set_transform(trans)
+        t.set_fontsize(2 * fontsize / 3)
+
+        return [t]
+
+
+class HandlerAnnotation(HandlerBase):
+    """
+    Handler for FancyArrowPatch instances.
```
Comment (afvincent) [inline lib/matplotlib/legend_handler.py]: FancyArrowPatch <- Annotation?

Comment (jlecoeur) [inline lib/matplotlib/legend_handler.py]: fixed!

**Comment by jlecoeur** [comment]
I forgot to ping you @bendichter @tacaswell 

I am not sure what should be displayed when an annotation has no text and no arrow. I used a white Rectangle but this still looks awkward, would it be better (and possible?) to remove the legend entry in this case?

**Comment by afvincent** [comment]
I wonder how useful it would be to make keyword arguments for the default replacement string for long text, and the maximal length of text entries (nice and clever names would have to be found…).  The use case I could see is a user that would want to deviate from the default values (e.g. from "a" for strings longer than 1 char to "Aa" for strings longer than 2 chars) without having to override the whole `create_artist` method of `HandlerText`. To do things similar to what is shown [here](http://matplotlib.org/devdocs/users/legend_guide.html#legend-handlers), i.e. (with dummy names for the kwargs)
```python
plt.legend(handler_map={my_annotation: HandlerAnnotation(rep_str='Aa', str_maxlen=2)})
```

**Comment by bendichter** [comment]
This looks beautiful! About the white rectangle, I can't think of a use case where you would have an annotation with no text and no arrow in the first place. I kind of liked having the white rectangle there because now I could imagine using it as a hack to give the legend subtitles.

**Comment by jlecoeur** [comment]
@bendichter Ok! I made the rectangle transparent to make sure it is not visible even if something is below the legend

@afvincent Good point, I added these arguments and updated the example in the first comment. I also prefer 'Aa', which conveys the idea of 'some text' better than 'a' so I made it the default.

**Comment by tacaswell** [comment]
@jlecoeur Any update on this?

**Comment by jlecoeur** [comment]
@tacaswell I would say it is done. Details like the relative size of character and arrow, or which character to use to represent long strings can be discussed, but the main thing is done and functional

**Comment by tacaswell** [comment]
"power cycled" to restart CI against current master.

**Comment by jlecoeur** [comment]
@tacaswell Many Ci fails 👎 I do not have time to fix them at the moment, can you have a shot at it? Otherwise I will come back to it, but that would not be before a month from now.

**Comment by tacaswell** [comment]
Look forward to hearing from you in a month!

**Comment by magedhennawy** [comment]
I'll take care of this for you!

**Comment by jklymak** [comment]
Closing in lieu of the active, but not merged, #10688

**Pull request closed without merging**